README.md
# myfavicon
Ajouter un favicon à votre blog


CHANGELOG
v0.8.1 23-04-2015 - Pierre Van Glabeke
* suppression icône + modif bloc licence

2013-12-30  Pierre Van Glabeke
* Compatibilité dc2.6

2008-04-21  Oleksandr Syenchuk  <sacha@xn--phnix-csa.net>
* Show a warning in HTML code if favicon mimetype is bad

2008-04-21  Oleksandr Syenchuk  <sacha@xn--phnix-csa.net>
* Formating changelog
* Erase cache directory during installation

2008-04-06  Oleksandr Syenchuk  <sacha@xn--phnix-csa.net>
* Option "Compatibilité Internet Explorer 6"
* Oubli d'une traduction : corrigé
 
2008-04-06  Oleksandr Syenchuk  <sacha@xn--phnix-csa.net>
* Traduction en français
 
2008-04-05  Oleksandr Syenchuk  <sacha@xn--phnix-csa.net>
* Un formulaire de configuration dans les préférences du blog est disponible
 
2008-04-05  Oleksandr Syenchuk  <sacha@xn--phnix-csa.net>
* La détection du type d'icône se fait automatiquement
 
2008-04-05  Oleksandr Syenchuk  <sacha@xn--phnix-csa.net>
* Première version paramètrable dans about:config


_admin.php
<?php 
/*
# This is 'My Favicon', a plugin for Dotclear 2
#
# Copyright (c) 2008-2015
# Oleksandr Syenchuk and contributors.
#
# This is an open source software, distributed under the GNU
# General Public License (version 2) terms and  conditions.
#
# You should have received a copy of the GNU General Public
# License along with 'My Favicon' (see COPYING.txt); 
# if not, write to the Free Software Foundation, Inc.,
# 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/
if (!defined('DC_CONTEXT_ADMIN')) { return; }

$core->addBehavior('adminBlogPreferencesHeaders', array('myFavicon', 'adminBlogPreferencesHeaders'));
$core->addBehavior('adminBlogPreferencesForm',array('myFavicon','adminBlogPreferencesForm'));
$core->addBehavior('adminBeforeBlogSettingsUpdate',array('myFavicon','adminBeforeBlogSettingsUpdate'));

class myFavicon
{
	public static function adminBlogPreferencesHeaders()
	{
		return '<script type="text/javascript" src="index.php?pf=myfavicon/blog_pref.js"></script>';
	}
	
	public static function adminBlogPreferencesForm($core,$settings)
	{
			$favicon_url = $settings->myfavicon->favicon_url;
			$favicon_ie_url = $settings->myfavicon->favicon_ie_url;

		echo
		'<div class="fieldset"><h4>Favicon</h4>'.
		'<p><label class="classic">'.
			form::checkbox('favicon_enable','1',(!empty($favicon_url) || !empty($favicon_ie_url))).
			__('Enable favicon').'</label></p>'.
		'<div id="favicon_config">'.
		'<p><label>'.__('Favicon URL for IE:').' '.
			form::field('favicon_ie_url',40,255,html::escapeHTML($favicon_ie_url)).'</label></p>'.
		'<p class="form-note warn">'
			.__('Please note, IE compatibility works only with ".ico" format.').'</p>'.
		'<p><label>'.__('Favicon URL:').' '.
			form::field('favicon_url',40,255,html::escapeHTML($favicon_url)).'</label></p>'.
		'</div></div>';
	}
	
	public static function adminBeforeBlogSettingsUpdate($settings)
	{
		$favicon_url = empty($_POST['favicon_url']) ? '' : $_POST['favicon_url'];
		$favicon_ie_url = empty($_POST['favicon_ie_url']) ? '' : $_POST['favicon_ie_url'];

			$settings->addNameSpace('myfavicon');
			$settings->myfavicon->put('favicon_url',$favicon_url);
			$settings->myfavicon->put('favicon_ie_url',$favicon_ie_url);
			$settings->addNameSpace('system');
	}
}

_define.php
<?php
/*
# This is 'My Favicon', a plugin for Dotclear 2
#
# Copyright (c) 2008-2015
# Oleksandr Syenchuk and contributors.
#
# This is an open source software, distributed under the GNU
# General Public License (version 2) terms and  conditions.
#
# You should have received a copy of the GNU General Public
# License along with 'My Favicon' (see COPYING.txt);
# if not, write to the Free Software Foundation, Inc.,
# 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/
if (!defined('DC_RC_PATH')) { return; }

$this->registerModule(
	/* Name */		"My Favicon",
	/* Description*/	"Add a customized favicon to your blog",
	/* Author */		"Oleksandr Syenchuk, Pierre Van Glabeke",
	/* Version */		'0.8.1',
	/* Properties */
	array(
		'permissions' => 'contentadmin',
		'type' => 'plugin',
		'dc_min' => '2.6',
		'support' => 'http://forum.dotclear.org/viewtopic.php?pid=323172#p323172',
		'details' => 'http://plugins.dotaddict.org/dc2/details/myfavicon'
	)
);

_public.php
<?php
/*
# This is 'My Favicon', a plugin for Dotclear 2
#
# Copyright (c) 2008-2015
# Oleksandr Syenchuk and contributors.
#
# This is an open source software, distributed under the GNU
# General Public License (version 2) terms and  conditions.
#
# You should have received a copy of the GNU General Public
# License along with 'My Favicon' (see COPYING.txt);
# if not, write to the Free Software Foundation, Inc.,
# 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/
if (!defined('DC_RC_PATH')) { return; }

if (is_callable('dcTemplate','SysBehavior')) {
	$core->addBehavior('publicHeadContent',array('myFavicon','publicHeadContent'));
} else {
	$core->addBehavior('templateBeforeValue',array('myFavicon','templateBeforeValue'));
}

class myFavicon
{
	#FIXME Mimetypes in common/lib.files.php (Clearbricks) are not enough
	public static $allowed_mimetypes = array(
		'ico' => 'image/x-icon',
		'png' => 'image/png',
		'bmp' => 'image/bmp',
		'gif' => 'image/gif',
		'jpg' => 'image/jpeg',
		'mng' => 'video/x-mng'
	);
	
	public static function publicHeadContent($core)
	{
		$res = self::faviconHTML($core->blog->settings);
		if (!empty($res)) {
			echo $res."\n";
		}
	}
	
	public static function templateBeforeValue($core,$id,$attr)
	{
		if ($id == 'include' && isset($attr['src']) && $attr['src'] == '_head.html') {
			return
			'<?php if (method_exists("myFavicon","faviconHTML")) {'.
			'echo myFavicon::faviconHTML();} ?>';
		}
	}

	private static function faviconHTML($settings)
	{
			$favicon_url = $settings->myfavicon->favicon_url;
			$favicon_ie_url = $settings->myfavicon->favicon_ie_url;

		if (empty($favicon_url) && empty($favicon_ie_url)) {
			return;
		}

		$ie_link = '';
		if ($favicon_ie_url != null) {
			$ie_link = '
<!--[if IE]><link rel="shortcut icon" type="image/x-icon" href="'.html::escapeHTML($favicon_ie_url).'" /><![endif]-->';
			
		}
		
		$other_link = '';
		if (!empty($favicon_url)) {
			$extension = files::getExtension($favicon_url);
			
			if (!isset(self::$allowed_mimetypes[$extension])) {
				$mimetype = files::getMimeType($favicon_url);
				if (!in_array($mimetype,self::$allowed_mimetypes)) {
					return '<!-- Bad favicon MIME type. -->'."\n";
				}
			}
			else {
				$mimetype = self::$allowed_mimetypes[$extension];
			}
			
			$other_link = '<link rel="icon" type="'.$mimetype.
			'" href="'.html::escapeHTML($favicon_url).'" />';
			
		}
			
		return $other_link.$ie_link;
	}
}

_install.php
<?php
/*
# This is 'My Favicon', a plugin for Dotclear 2
#
# Copyright (c) 2008-2015
# Oleksandr Syenchuk and contributors.
#
# This is an open source software, distributed under the GNU
# General Public License (version 2) terms and  conditions.
#
# You should have received a copy of the GNU General Public
# License along with 'My Favicon' (see COPYING.txt);
# if not, write to the Free Software Foundation, Inc.,
# 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/
if (!defined('DC_CONTEXT_ADMIN')) { return; }

$m_version = $core->plugins->moduleInfo('myfavicon','version');
$i_version = $core->getVersion('myfavicon');

if (version_compare($i_version,$m_version,'>=')) {
	return;
}

# Création du setting (s'il existe, il ne sera pas écrasé)
$settings = new dcSettings($core,null);
$settings->addNamespace('myfavicon');
$settings->myfavicon->put('favicon_url','','string','Favicon URL',false);
$settings->myfavicon->put('favicon_ie_url','','string','Favicon URL Internet Explorer',false);

# La procédure d'installation commence vraiment là
$core->setVersion('myfavicon',$m_version);